federationsender/queue/destinationqueue.go

```diff
@@ -51,59 +53,62 @@ func (oq *destinationQueue) sendEvent(ev *gomatrixserverlib.HeaderedEvent) {
 	oq.runningMutex.Lock()
 	defer oq.runningMutex.Unlock()
 	oq.pendingEvents = append(oq.pendingEvents, ev)
-	if !oq.running {
-		oq.running = true
+	if !oq.running.Load() {
 		go oq.backgroundSend()
 	}
 }
 
 // sendEDU adds the EDU event to the pending queue for the destination.
 // If the queue is empty then it starts a background goroutine to
-// start sending event to that destination.
+// start sending events to that destination.
 func (oq *destinationQueue) sendEDU(e *gomatrixserverlib.EDU) {
 	oq.runningMutex.Lock()
 	defer oq.runningMutex.Unlock()
 	oq.pendingEDUs = append(oq.pendingEDUs, e)
-	if !oq.running {
-		oq.running = true
+	if !oq.running.Load() {
+		go oq.backgroundSend()
+	}
+}
+
+// sendInvite adds the invite event to the pending queue for the
+// destination. If the queue is empty then it starts a background
+// goroutine to start sending events to that destination.
+func (oq *destinationQueue) sendInvite(ev *gomatrixserverlib.InviteV2Request) {
+	oq.runningMutex.Lock()
+	defer oq.runningMutex.Unlock()
+	oq.pendingInvites = append(oq.pendingInvites, ev)
+	if !oq.running.Load() {
 		go oq.backgroundSend()
 	}
 }
 
+// backgroundSend is the worker goroutine for sending events.
 func (oq *destinationQueue) backgroundSend() {
+	oq.running.Store(true)
+	defer oq.running.Store(false)
+
 	for {
-		t := oq.next()
-		if t == nil {
+		transaction, invites := oq.nextTransaction(), oq.nextInvites()
+		if !transaction && !invites {
```

Comment:
This feels like this should be possible and more easily read if it used channels instead of mutexes and shared memory. Not a task for now but thoughts welcome.

Reply:
I actually almost did this but I figured it might end up spiralling the PR a bit further than necessary. I also figured that it would need some thought on channel buffer size vs what might happen if a channel send blocks for some reason. I'll create a separate issue for looking into this!

Reply:
Thanks!